text/0000-macro-expansion-for-macro-input.md

```diff
@@ -0,0 +1,214 @@
+
+- In a minimal implementation we want to keep expansion result interpolation as simple as possible - this means avoiding enqueuing an expansion that expands _inside of_ another enqueued expansion.
+- One solution is to recursively parse the input of marked macros until we find a marked macro with none in its input, adding only this innermost call to the expansion queue.
+
+    This increases the amount of re-parsing (since after every expansion we're repeatedly parsing macro bodies looking for innermost calls) at the cost of fewer re-expansions (since each marked macro will only ever see its input after all marked macros have been expanded).
+
+If a marked `#` token is part of an inner-attribute `#![...]` call, the situation is similar: the parser can forward-parse the macro path and token arguments, and with a little work can forward-parse the body.
+
+When the parser encounters a marked `!` token, it needs to forward-parse the token arguments, but also needs to _backtrack_ to parse the macro path. In a structured area of the grammar (such as in an attribute macro body or a structured decl macro) this would be fine, since we would already be parsing an expression or item and hence have the path ready. In an _unstructured_ area we would actually have to backtrack within the token stream and 'reverse parse' a path: is this an issue?
+
+## Delayed resolution of unresolved macros
+
+The existing expansion loop adds any currently unresolved macros to a _resolution_ queue. When re-parsing macro output, if any newly defined macros would allow those unresolved macros to be resolved, they get added to the current expansion queue. If there are unresolved macros but no macros to expand, the compiler reports the unresolvable definition.
+
+The new expansion order described [above](#macro-expansion-and-marking) is designed to expand all marked macros as much as possible before trying to expand unmarked ones. We know that marked macros are always in token position, so expansion-eligible unmarked macros are the only way to introduce new macro definitions.
+
+In the new order, we still accumulate unresolved macros (marked and unmarked), and we still remove them from the resolution queue to the relevant expansion queue whenever they get defined. The only difference is an extra error case, where a resolved unmarked macro has an unresolved marked macro in its input, and there are no unmarked macros to expand. In this case, the resolution queue still contains the unresolved marked macro, and so the compiler again reports the unresolvable definition.
+
+# Drawbacks
+
+This proposal:
+
+* Relies on proc macro authors doing macro expansion. This might partition the macro ecosystem into expansion-ignoring (where input macro calls are essentially forbidden for any part of the input that needs to be inspected) and expansion-handling (where they work fine _as long as_ the proc macro author has used the expansion API correctly).
```

Comment:
I don't think that this is a drawback: Currently it's just not possible to ignore expansion if you want to do anything to bang-macro arguments, so in a sense *all* current procedural macros are expansion-ignoring, because they have no alternative. For those macros that can fulfil their task while ignoring expansion, nothing will change, whereas the proposed implementation makes it possible to look into expansions at all.

Reply:
Right, newer macros that care about expansion can repeatedly "drive" any expansion-ignorant macros they encounter to completion, so I guess I can trim this.